ext/rugged/rugged_tree.c

```diff
@@ -93,6 +93,63 @@ static VALUE rb_git_tree_entrycount(VALUE self)
 	return INT2FIX(git_tree_entrycount(tree));
 }
 
+static int rugged__treecount_cb(const char *root, const git_tree_entry *entry, void *data)
+{
+	struct rugged_treecount_cb_payload *payload = data;
+
+	if (payload->limit >= 0 && payload->count >= payload->limit)
+		return -1;
+	else if(git_tree_entry_type(entry) == GIT_OBJ_TREE)
+		return 0;
+	else {
+		++(payload->count);
+		return 1;
+	}
+}
+
+/*
+ *  call-seq:
+ *    tree.count_recursive(limit=nil) -> count
+ *
+ *  `limit` - The maximum number of blobs to the count in the repository.
+ *  Rugged will stop walking the tree after `limit` items to avoid long
+ *  execution times.
+ *
+ *  Return the number of blobs (up to the limit) contained in the tree and
+ *  all subtrees.
+ */
+static VALUE rb_git_tree_entrycount_recursive(int argc, VALUE* argv, VALUE self)
+{
+	git_tree *tree;
+	Data_Get_Struct(self, git_tree, tree);
+	int error;
+	struct rugged_treecount_cb_payload payload;
+	VALUE rbLimit;
+
+	if (rb_scan_args(argc, argv, "01", &rbLimit) == 0)
+		rbLimit = Qnil;
+
+	if (NIL_P(rbLimit))
+		payload.limit = -1;
+	else if (TYPE(rbLimit) != T_FIXNUM)
+		rb_raise(rb_eTypeError, "limit must be a Fixnum");
+	else
+		payload.limit = FIX2INT(rbLimit);
```

Comment:
this could maybe be simplified to something like:

``` C
if (!NIL_P(rbLimit)) {
  Check_Type(rbLimit, T_FIXNUM);
  payload.limit = FIX2INT(rbLimit);
}
```

`Check_Type` will raise an `ArgumentError` for you.